collect_app/src/test/java/org/odk/collect/android/utilities/CustomTabHelperTest.java

```diff
@@ -0,0 +1,33 @@
+package org.odk.collect.android.utilities;
+
+import android.net.Uri;
+
+import androidx.test.core.app.ApplicationProvider;
+
+import junit.framework.TestCase;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+@RunWith(RobolectricTestRunner.class)
+public class CustomTabHelperTest extends TestCase {
+
+    @Test
+    public void uriShouldBeNormalized() {
+        CustomTabHelper customTabHelper = spy(new CustomTabHelper());
+        doNothing().when(customTabHelper).openUriInChromeTabs(any(), any());
+        doNothing().when(customTabHelper).openUriInWebView(any(), any());
+        doNothing().when(customTabHelper).openUriInExternalBrowser(any(), any());
+
+        Uri uri = mock(Uri.class);
+        customTabHelper.openUri(ApplicationProvider.getApplicationContext(), uri);
+        verify(uri).normalizeScheme();
```

Comment:
This verifies that `normalizedScheme()` is called but there's no guarantee that we'll use the result of that call in our `ACTION_VIEW` intent. So here I could change `uri = uri.normalizeScheme();` in `CustomTabHelper` to just `uri.normalizeScheme();` and the test would still pass. 

We could test this without the `spy` by passing in a Uri that would get normalized (like `HTTP://www.android.com`) and then verifying that we see a normalized version in the resulting intent (using `shadowOf(application).getNextStartedActivity()`). The complexity here is that we'd need to have the `openUriInChromeTabs` route fail, so we fall through into the path we want to check. For this we'd need to pull out a component to represent that we could set up to fail.

That all seems like a lot of work, but I'd generally say that if something is hard to test it's a sign it needs rework. I'm also not clear on what's happening with the Chrome tab stuff here and why we're doing it.

All that said this is a crash fix, so it'd be nice to get into the release, so I don't want my pretentiousness to get in the way too much 😉. Maybe we open another PR for the test, and we can discuss/tweak there?

Reply:
Right it's a silly test but yeah maybe we can improve it later in a separate pr.